### PR TITLE
Mp4muxer tests

### PR DESCRIPTION
**Pull request by zhuker**

### CODE REVIEW

File: src/main/java/org/jcodec/common/io/ByteBufferChannel.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * This class is part of JCodec ( www.jcodec.org ) This software is distributed
+ * under FreeBSD License
+ * 
+ * Implements a seekable byte channel that wraps a byte buffer
+ * 
+ * @author The JCodec project
+ * 
+ */
+public class ByteBufferChannel implements SeekableByteChannel {
```
Comment (svitvitskiy) [inline src/main/java/org/jcodec/common/io/ByteBufferChannel.java]: Is this different from ByteBufferSeekableByteChannel ? Can we delete one?

File: src/main/java/org/jcodec/containers/mp4/muxer/CodecMP4MuxerTrack.java [hunk 2]
```diff
@@ -117,14 +118,11 @@ public void addFrameInternal(Packet pkt, int entryNo) throws IOException {
...
-            if (codec == Codec.H264) {
+            if (codec == Codec.H264 && !spsList.isEmpty()) {
                 SeqParameterSet sps = SeqParameterSet.read(spsList.get(0).duplicate());
                 Size size = H264Utils.getPicSize(sps);
-                VideoCodecMeta meta = org.jcodec.common.VideoCodecMeta.createSimpleVideoCodecMeta(size,
-                        ColorSpace.YUV420);
+                VideoCodecMeta meta = createSimpleVideoCodecMeta(size, ColorSpace.YUV420);
                 addVideoSampleEntry(meta);
-            } else {
-                checkState(false, "Sample entry missing not supported for anything other then H.264");
```
Comment (svitvitskiy) [inline src/main/java/org/jcodec/containers/mp4/muxer/CodecMP4MuxerTrack.java]: I would convert this to a warning, i.e. Logger.warn("Creating a track without sample entry");

File: src/main/java/org/jcodec/containers/mp4/muxer/CodecMP4MuxerTrack.java [hunk 3]
```diff
@@ -153,9 +151,15 @@ void addVideoSampleEntry(VideoCodecMeta meta) {
 
     public void setCodecPrivateIfNeeded() {
         if (codec == Codec.H264) {
-            getEntries().get(0).add(H264Utils.createAvcCFromPS(selectUnique(spsList), selectUnique(ppsList), 4));
+            List<ByteBuffer> sps = selectUnique(spsList);
+            List<ByteBuffer> pps = selectUnique(ppsList);
+            if (!sps.isEmpty() && !pps.isEmpty()) {
+                getEntries().get(0).add(H264Utils.createAvcCFromPS(sps, pps, 4));
+            }
```
Comment (svitvitskiy) [inline src/main/java/org/jcodec/containers/mp4/muxer/CodecMP4MuxerTrack.java]: I would also add a warning here:
} else {
  Logger.warn("Not adding a sample entry for h.264 track, missing any SPS/PPS NAL units");
}

File: src/main/java/org/jcodec/containers/mp4/muxer/CodecMP4MuxerTrack.java [hunk 4]
```diff
@@ -153,9 +151,15 @@ void addVideoSampleEntry(VideoCodecMeta meta) {
...
         } else if (codec == Codec.AAC) {
-            getEntries().get(0).add(EsdsBox.fromADTS(adtsHeader));
+            if (adtsHeader != null) {
+                getEntries().get(0).add(EsdsBox.fromADTS(adtsHeader));
+            }
```
Comment (svitvitskiy) [inline src/main/java/org/jcodec/containers/mp4/muxer/CodecMP4MuxerTrack.java]: I would also add a warning:
I would also add a warning here:
} else {
  Logger.warn("Not adding a sample entry for AAC track, missing any ADTS headers.");
}

File: src/main/java/org/jcodec/common/io/ByteBufferChannel.java (hunk 1 above)
Comment (zhuker) [inline src/main/java/org/jcodec/common/io/ByteBufferChannel.java]: they work in slightly different way
one is basically for writing another one for reading
i guess i should add testcases

Comment (zhuker) [inline src/main/java/org/jcodec/common/io/ByteBufferChannel.java]: see https://github.com/jcodec/jcodec/commit/a19f54ddaf2c6a69bfb3e9ba2c26fca189aacf19

**Comment by zhuker** [comment]
@svitvitskiy see updates

**Pull request merged**